libnavigation-core/src/test/java/com/mapbox/navigation/core/MapboxNavigationRerouteInterruptionTest.kt

```diff
@@ -0,0 +1,91 @@
+package com.mapbox.navigation.core
+
+import com.mapbox.navigation.base.ExperimentalPreviewMapboxNavigationAPI
+import com.mapbox.navigation.core.reroute.NavigationRerouteController
+import com.mapbox.navigation.core.reroute.RerouteController
+import com.mapbox.navigation.core.reroute.RerouteState
+import io.mockk.every
+import io.mockk.mockk
+import io.mockk.verify
+import kotlinx.coroutines.ExperimentalCoroutinesApi
+import kotlinx.coroutines.InternalCoroutinesApi
+import org.junit.Before
+import org.junit.Test
+import org.junit.runner.RunWith
+import org.robolectric.ParameterizedRobolectricTestRunner
+import org.robolectric.annotation.Config
+
+@ExperimentalPreviewMapboxNavigationAPI
+@Config(shadows = [ShadowReachabilityFactory::class])
+@InternalCoroutinesApi
+@ExperimentalCoroutinesApi
+@RunWith(ParameterizedRobolectricTestRunner::class)
+internal class MapboxNavigationRerouteInterruptionTest(
+    private val previousState: RerouteState?,
+    private val expectedNumberOfInterruptions: Int,
+) : MapboxNavigationBaseTest() {
+
+    companion object {
+
+        @JvmStatic
+        @ParameterizedRobolectricTestRunner.Parameters
+        fun data(): List<Array<Any?>> {
+            return listOf<Array<Any?>>(
+                arrayOf(null, 0),
+                arrayOf(RerouteState.Idle, 0),
+                arrayOf(RerouteState.Interrupted, 0),
+                arrayOf(RerouteState.FetchingRoute, 1),
+                arrayOf(mockk<RerouteState.Failed>(), 0),
```

Comment:
Maybe?
```suggestion
                arrayOf(RerouteState.Failed("test"), 0),
```

Reply:
Why? I don't need any details about the state in my test, only its class.

Reply:
> How often did you have a situation when bunch of tests failed after refactoring/improvements just because mock didn't behave like a real object for a new case?

It happens, However, when I didn't use mocks, I used to run into a situation where half the project was warmed up and initialised  all sorts of things, including some static stuff because of which some random future tests failed and it was hell to try and figure out what went wrong.
And for lots of situations that you describe, if a test fails because some method is not mocked, it's probably an indication that you forgot to test something. Because otherwise you would have mocked this behaviour.

Reply:
>  when I didn't use mocks, I used to run into a situation where half the project was warmed up and initialised all sorts of things, including some static stuff because of which some random future tests failed and it was hell to try and figure out what went wrong.

That's a terrible experience 😱. I don't want this to happen in our project. 
Do you think it's going to happen every time you use real objects instead of mocks?

Reply:
Depends on the object of course. If it's a simple data class it's probably safe. However, there could be some default arguments that use some static stuff to retrieve the default values and so on. One has to be very attentive.
But I don't see lots of problems with mocks. In most cases they return default values if we use `relaxed = true` and if we didn't change the behaviour in test, we probably don't care about it and default is fine. If it breaks something, that's a good indication that we might wanna add more tests.

Reply:
For example, you have a test that uses a mock, let's say we mock class A. And the test breaks when A added some properties and methods that are not mocked in this test. Why did the test fail? Probably because our class that we test uses this new API from A. Why didn't we mock it? Probably we forgot to add tests for this behaviour.
I see it this way, but maybe I didn't understand the problem you were talking about correctly.

Reply:
That's a long discussion which has already started in your PR and continued in the call. We think a bit differently. :)
How critical do you find it here? Can I merge?